Make dedicated fiber-photometry session json generation

dougollerenshaw opened the pull request:
# Simplify and Improve Fiber Photometry ETL Implementation

## Motivation
The previous implementation was very specific to one experimental paradigm where fiber photometry was combined with optogenetic stimulation. We want a general fiber photometry session generator which can be paired with any other experimental platform. This PR accomplishes that by creating a more modular and focused implementation.

## Key Changes

### Naming and Organization
- Created a dedicated `utils.py` in the `fip` module to house helper functions, improving code organization and maintainability
- Moved time extraction and file handling logic to `utils.py`, keeping the ETL implementation clean and focused

### Core Functionality
- Removed all references to optogenetic stimulation experiment in `fip/session.py`
- Implemented automatic session time extraction through:
  - Start time: Parsed from data file names (format: `FIP_DataG_YYYY-MM-DDThh_mm_ss.csv`)
  - End time: Calculated from timestamps in CSV data files
  - Proper timezone handling for consistent UTC conversion
- Simplified the ETL interface to focus solely on fiber photometry data processing

### Documentation
- Added comprehensive README ([direct link](https://github.com/AllenNeuralDynamics/aind-metadata-mapper/blob/feat-250-update-fiber-metadata/src/aind_metadata_mapper/fip/README.md)) with:
  - Clear usage examples
  - Command-line interface documentation
  - Explanation of automatic time extraction
  - Description of expected file formats and naming conventions

### Testing
- Maintained 100% test coverage for FIP module
- Added specific tests for:
  - Time extraction utilities
  - File handling functions
  - Timezone conversion edge cases
  - Error handling scenarios

## Impact
These changes make the fiber photometry ETL more:
- Modular: Can be easily integrated with other experimental platforms
- Maintainable: Clear separation of concerns between ETL and utilities
- Robust: Better error handling and time zone management
- User-friendly: Improved documentation and examples

## Important note on standards/file conventions
This PR makes a number of implicit assumptions about the file naming conventions and the file structures as outlined in the v0.2.2 of the file standards doc [here](https://github.com/AllenNeuralDynamics/aind-file-standards/blob/313f1afc1eef7512335c2c494c952268dad9bd28/file_formats/fip.md). If those naming conventions change, this mapper will cease to work. Ideally we would have a more structured way of defining and documenting these conventions to avoid confusion.

Closes #250

arielleleon commented on .github/workflows/run_main_tests.yml:
```diff
@@ -22,9 +22,15 @@ jobs:
     - name: Run linter checks
-      run: flake8 . && interrogate --verbose .
+      run: |
+        echo "=== Running style checks with flake8 ==="
```
Can you ask Jon if he can add these prints to the template?

dougollerenshaw replied:
Thanks @arielleleon for the review! Agreed that these print statements would be useful in the template and other repos. I was motivated to add this because I was confused about the coverage report generated by the `interrogate --verbose .` I interpreted it as a test coverage report, when it was actually a docstring coverage report. So I spent quite a long time (more than I want to admit!) trying to understand why I had 100% test coverage locally but couldn't get past this check. 

@jtyoung84 , tagging you for visibility. I'm about to request review from you also.

dougollerenshaw commented on src/aind_metadata_mapper/fib/models.py:
```diff
@@ -0,0 +1,35 @@
+    mouse_platform_name: str
+    active_mouse_platform: bool
+    data_streams: List[dict]
+    session_type: str = "Fiber_Photometry"
```
Good point. Yes, it probably should. See https://github.com/AllenNeuralDynamics/aind-metadata-mapper/pull/251/commits/de7a1d9a333f8d2ed7805d51c4f6b967228e7a4c. 

One complication, though, is that when this is joined with the corresponding behavior metadata, we'll have to figure out how to reconcile the session types

dougollerenshaw commented:
Important note here. The exposure time in the data streams is wrong. See https://github.com/AllenNeuralDynamics/aind-data-migration-scripts/issues/355. I'll need to be sure to populate with correct values from @hagikent before merging.

jtyoung84 commented on src/aind_metadata_mapper/fib/utils.py:
```diff
@@ -0,0 +1,227 @@
+import json
+
+
+def convert_ms_since_midnight_to_datetime(
```
Timezone handling is notoriously tricky. There are many edge cases that need to be tested. In general, it's best to convert everything to timestamps or UTC. For example, python's datetime object doesn't handle daylight savings time very well. Assuming that the ms_since_midnight are the actual milliseconds elapsed since midnight of the local time, then converting the local time time first and then applying the time delta circumvents that edge case:
```python
from datetime import datetime, timedelta, time
from zoneinfo import ZoneInfo
from aind_metadata_mapper.fib.utils import convert_ms_since_midnight_to_datetime

def convert_ms(ms_since_midnight, base_date):
    base_date_midnight = datetime.combine(base_date.date(), datetime.min.time())
    base_date_midnight_utc = base_date_midnight.astimezone(ZoneInfo("UTC"))
    delta = timedelta(milliseconds=ms_since_midnight)
    return base_date_midnight_utc + delta

base_date = datetime(2025, 3, 9, 0, 0, 0, tzinfo=ZoneInfo(key='America/Los_Angeles'))

# These output the same datetime object even though the millis are different
convert_ms_since_midnight_to_datetime(1000*60*60*2.5, base_date)
# datetime.datetime(2025, 3, 9, 10, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))

convert_ms_since_midnight_to_datetime(1000*60*60*3.5, base_date)
# datetime.datetime(2025, 3, 9, 10, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))

# These output different objects
convert_ms(1000*60*60*2.5, base_date)
# datetime.datetime(2025, 3, 9, 10, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))

convert_ms(1000*60*60*3.5, base_date)
# datetime.datetime(2025, 3, 9, 11, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))
```

dougollerenshaw replied:
Good point. I updated the function as you suggested and also added specific test cases for March 9, 2025 (the day we switched to DST in 2025). See https://github.com/AllenNeuralDynamics/aind-metadata-mapper/pull/251/commits/6ba5ba9dc9b55e32ac99522ee4019ab81468b811

jtyoung84 reviewed (commented):
I left one note about the time zone handling that may need to be resolved. I'll leave a few comments without requesting changes:

- We've been using NumPy's docstring format for our docstring formats
- We discourage writing temp files during unit testing. We usually mock the write and compare things in-memory
- We encourage large PRs to be broken up into ~500 line changes if possible

jtyoung84 reviewed (changes requested):
The `convert_ms_since_midnight_to_datetime` may need to be updated to handle edge cases.

jtyoung84 commented on src/aind_metadata_mapper/fib/session.py:
```diff
@@ -0,0 +1,295 @@
+        return self._load(transformed, self.job_settings.output_directory)
+
+
+def update_job_settings_with_times(
```
It seems like this method should either be added to the utils package or be added as a class method.

dougollerenshaw replied:
I removed this function entirely.

jtyoung84 commented on src/aind_metadata_mapper/fib/session.py:
```diff
@@ -0,0 +1,295 @@
+            job_settings = JobSettings(**json.loads(job_settings))
+        super().__init__(job_settings)
+
+    def _extract(self) -> JobSettings:
```
The extract methods usually parse the input sources into intermediate data models that can be transformed into the output Session object.

dougollerenshaw replied:
I made some heavy modifications to session.py. Let me know if that satisfies you.

jtyoung84 commented on src/aind_metadata_mapper/fib/session.py:
```diff
@@ -0,0 +1,295 @@
+
+        return settings
+
+    def _create_stream(
```
This method could be decorated as a `@staticmethod` and the `self` arg can be removed. Alternatively, the settings could potentially be accessed directly via `self.job_settings`.

dougollerenshaw replied:
Removed entirely.

hagikent reviewed (approved):
main functions worked

dougollerenshaw commented:
@jtyoung84:

> I left one note about the time zone handling that may need to be resolved. I'll leave a few comments without requesting changes:

Timezone fix implemented.

> 
> We've been using NumPy's docstring format for our docstring formats

fixed

> We discourage writing temp files during unit testing. We usually mock the write and compare things in-memory

fixed. using mock now

> We encourage large PRs to be broken up into ~500 line changes if possible

One reason for the big diff was that I had renamed the module from "fip" to "fib" to be consistent. So there were a lot of full file deletions/additions. But I reverted those changes to keep the diff more manageable. I'm still really bothered by this naming convention mismatch, though, and think it should be addressed in a followup PR. It's complicated because the scientific team uses "FIP" but the SciComp standard seems to be "FIB" (e.g. we require that fiber data be in a folder named "fib"). So we should probably work to get everyone to agree on one or the other and change it everywhere to match!
